initial API doc

rdallman opened the pull request:
this does not track the current API, but is an attempt to document what we
want the API to look like to then go make it happen. it's easier to do now
that we have some data, unsurprisingly.

aside from notes in TODO, some notes on changes:

* remove CMCPrice and totalSupply from token (helps cachability, looks unused, moved to stats)
* remove token object in get pair, just give address for each (up for debate, wanna see usage?)

TODO to dig around the uniswap interface to see about anything we may want to
add that isn't covered well here

rdallman committed:
initial API doc

this does not track the current API, but is an attempt to document what we
want the API to look like to then go make it happen. it's easier to do now
that we have some data, unsurprisingly.

aside from notes in TODO, some notes on changes:

* remove CMCPrice and totalSupply from token (helps cachability, looks unused, moved to stats)
* remove token object in get pair, just give address for each (up for debate, wanna see usage?)

TODO to dig around the uniswap interface to see about anything we may want to
add that isn't covered well here

treeder commented:
closes #6

treeder commented on api.md:
````diff
@@ -0,0 +1,242 @@
+
+```
+{
+  "stats": [
````
Should the results be separated by the tokens?  Or could remove this endpoint and only have the stats by token or pair. Or perhaps in your comment this only returns sums. But maybe this endpoint should be merged into /v1/tokens or /v1/pairs (sort of like it is now), just with the stats data as fields on the Pair/Token. It seems with this proposal you'd have to call /v1/tokens to get the token data, then /v1/stats/tokens to get the summed up data, then merge those yourself.

rdallman replied:
yea I agree it's weird to have histograms across all tokens interspersed, thanks, will get rid of it. after thinking about it, think we just remove 'interval' and force sums for that endpoint between given start and end, so it's easy to do things like 7d or 24h. really wanted a 'cheap' endpoint just to list tokens/pairs, it's kinda standard api thing I guess, I know primary use case atm is just to get addresses to use but still - I follow on merging the two, it's a little kludgy I guess but if you really just want to get a name/address for a token it's a lot of TMI (as API stands, no way to get name / symbol of token without having handle on its address other than grabbing it from list endpoint - could change that too). in a way the 'list all' is kind of just an optimization for looking up a token details by its address to fill in (N requests), to my eye anyway. 

I think this endpoint still may need a little sugar to specify a subset of tokens possibly but it can wait (eg `?in=ETH,WGO,YFI` for just getting one page of a list or something along those lines, can wait for use case tho).

yeah, flow wise I was thinking we need a few things really for a token or pair:

* list all
* get volume sums of 24h, 7d, etc
* get latest liquidity (currently merged into ^)
* get latest price (token)
* get fee sums of 24h, 7d, etc
* get histogram of volume / price

so tried to logically divide this up in the same manner. I think each column being a request (staring at uniswap page here) is kinda what I'm going for, or on token / pair detail page need histograms on top of latest numbers but a request for each is still pretty good I think. if you take this a step further you get `/stats/tokens/{address}/volume` and things like that, to allow more specificity basically. but, makes my head spin a bit thinking about caching all of this. anyway, sums!

treeder replied:
What if we add a param to include stats on /v1/pairs and /v1/tokens to include sums/liquidity? Like ?stats=24h would return stats, without it it would just return the static info.

rdallman replied:
yea, that could work, makes me want to rethink api tho (ie why have `/stats` at all then?). also pretty odd in the backend itself, have to create a new struct and merge the two and return, so gains seem kind of minimal in that sense, just picking to do it in the backend instead of the frontend at that point I think. kinda per comment, hoping to map each endpoint to 1 db call to start out, not least because it makes it super easy to cache, main benefit is just simplicity. guessing at some point we end up doing more complicated stats reports but not sure how that will bubble up either, api is flexible to that extent. only concern i'm tinkering with having the sums separate is whether to have `map[address]stat{}` in the json but it's also kind of just a convenience, in my head the front end will be something like this for the 'sums' page:

* get /tokens -> add to map[address]token (same idea for pairs), probably tote this around for various pages
* get token sums/liquidity/price -> sort by favorite column (eg volume), do quick lookup in tokens map from ^ to get token symbol to display instead of the address when rendering each row, et voila?

this reminds me I need to average price for the sums endpoint, or could use latest like liquidity. I guess either one is interesting, it's possible now to get buckets and get average for anyone interested, so nvm (punt).

rdallman replied:
oh nvm we have pair and symbol in the respective pair and token buckets now, so the previous comment is moot, originally only pair buckets had pair which I was gonna bring up but this seems uniform now and I think solves the rift here? I think having them in the bucket works well enough, at worst it wastes a few (very few) bytes but it's not so big a deal. if we wanted to show full description (maybe for hover I guess?) it's probably only gonna be on token/pair detail page I think so that gets cheaper. anyway, proceeding with sums.

treeder reviewed (commented):
This seems good. The one concern would be

rdallman commented on api.md:
```diff
@@ -0,0 +1,242 @@
+apart, between `time_start` and `time_end`.
+
+TODO defaults for `time_` fields
+TODO do we want string encapsulated floats?
```
any thoughts on this @treeder ? it's a bit odd, and I think it's simply for decimal.Decimal package (ie it's the default), which notes things about javascript having issues losing precision when parsing them. I'm not sure if this is intentional ("a feature"), as I simply don't have any context for why we didn't use floats vs the decimal package (I do see the firebase oddness, unsure if related).

treeder replied:
If we care about precision, then yes, we need to stringify them. If we don't really care, then it could be floats. For stats, we probably don't care 100%. But I could foresee that it might matter in certain cases, so maybe best to just use strings across the board so we don't end up with some floats and some strings.

rdallman committed:
update with latest discussion

rdallman committed:
add defaults for times

rdallman commented:
for the collect endpoint, is this just for a forced collection? if this thing runs as a daemon it should be relatively easy to just have it run on a schedule without needing the endpoint called, and we can pretty easily put a little lock in the db for multiple daemons to only let one run the actual collection. thoughts?

rdallman committed:
first pass v1 api

I haven't tested this at all, and is mostly a pass in a hurry, so kinda TODO
on all of this, but this is the basic idea I think, it makes the front end a
lot simpler imo (I mean, some of the methods, specifically) and just maps to
the backend which can be more straight forward now.

rdallman commented:
I made a first pass, it's not really ready to get picked over with a fine tooth comb for bugs or anything, but if this seems to be going down a wrong path please let me know... pushing in case osx overnight update bomb mostly 🤞

rdallman committed:
change cache default

rdallman committed:
fix pairs sums, totals api

pair sums w/o address specified were all getting mashed together, fix this
from backend.

TODO need to do same for tokens
TODO add tests!

rdallman committed:
fix up tokens sums, api doc required fields and sorts

rdallman committed:
add testing facilities

TODO add errors, validation & require time/frame stuff

rdallman commented:
made good progress, got testing facilities set up and shoved the 'leaderboard' down (sort is removed, for now, thinking about it) and got it working so it can just come straight from cache (no memory overhead, cpu, etc), tomorrow can start adding actual tests and errors / validation. checked in the checker which was handy, gonna solidify in real tests...

rdallman committed:
got first couple tests in

rdallman committed:
fix liquidity blank, change uri variable names to address

treeder commented:
Is this a breaking change? ie: will UI need to update to align?

rdallman commented:
yes, will need a small update, mostly cosmetic stuff I believe.

rdallman committed:
passing totals test

rdallman committed:
go back to old now tested ways

rdallman committed:
add checker again?

rdallman committed:
match pair api to doc

treeder commented:
> for the collect endpoint, is this just for a forced collection?

It's on a schedule, it's already running so don't change how that part works.

rdallman committed:
match api doc for query params

rdallman committed:
add back default liquidity sort TODO need to update docs and add query parameter

rdallman committed:
require time_start, time_end; add sort facilities

see docs for sorting, it's pretty simple. the default is -liquidityUSD for the
endpoints where it was. we can change the semantics, the popular opinion seems
to be `[+|-]key` or `key[:asc|:desc]` api wise, but open to ideas. it ended up
being a little messy to allow sorting any field, but it seems to work. trying
to read into firebase docs more to see if we can push it all down, but it
doesn't seem easy unless we only have collections for hourly and daily and
then can query against those directly (ie no client side aggregation)

rdallman commented:
alright i got changes in to required the time_start and time_end fields now or error and added ability to sort by any field for the token / pair sum api. can add to the single pair/token api too (it's just chronological for now). there's a future hitch for the sorting when we add paging (firebase woes) but the api will probably be fine as it is so we can punt on that. in any case, got most of the nagging todo done now, can pick when to cut over.

rdallman committed:
update checker

rdallman committed:
fix up windows to go reverse chronological

this is more likely to align windows as end is more likely to be a time
definitely within our system whereas start is the beginning of time, walking
forward from start was leading to window mismatches when 24h data was compared
to 60 day @24h data that were both displaying varying 24h intervals that did
not align. now they align with 'end' time given and go backwards. the cost of
this is order n to go back over the times in reverse order to do the sums
here, though we could add an index for descending time and not need to absorb
this cost (master, effectively, has this cost already, so we're even but
still), i think this is kind of a hot fix and a more proper fix is to limit
the domain for intervals and aggregate and store only daily and hourly
intervals, which will fix the daily data window mismatch thing (eg we were
comparing 1am yesterday-1am today vs 4am yesterday-4am today, and these didn't
match most often).

in any case, agree this is gross / sucks but it does fix the issue.

rdallman committed:
make a note

rdallman committed:
fix sort bug

rdallman committed:
align json delimiter oops
